src/zx_layout.jl

```diff
@@ -7,23 +7,72 @@ import Base: copy
 A struct for the layout information of `ZXDiagram` and `ZXGraph`.
 """
 struct ZXLayout{T<:Integer}
-    nbits::T
-    spider_seq::Vector{Vector{T}}
+    nbits::Int
+    spider_q::Dict{T, Rational{Int}}
+    spider_col::Dict{T, Rational{Int}}
 end
 
-ZXLayout(nbits::T, spider_seq::Vector{Vector{T}}) where {T} = ZXLayout{T}(nbits, spider_seq)
-ZXLayout{T}() where {T} = ZXLayout(T(0), Vector{T}[])
+ZXLayout(nbits::Integer, spider_q::Dict{T, Rational{Int}}, spider_col::Dict{T, Rational{Int}}) where {T} = ZXLayout{T}(Int(nbits), spider_q, spider_col)
+ZXLayout{T}() where {T} = ZXLayout(0, Dict{T, Rational{Int}}(), Dict{T, Rational{Int}}())
 
-copy(layout::ZXLayout) = ZXLayout(copy(layout.nbits), deepcopy(layout.spider_seq))
+copy(layout::ZXLayout) = ZXLayout(layout.nbits, deepcopy(layout.spider_q), deepcopy(layout.spider_col))
 function rem_vertex!(layout::ZXLayout{T}, v::T) where {T}
-    for seq in layout.spider_seq
-        deleteat!(seq, seq .== v)
-    end
+    delete!(layout.spider_q, v)
+    delete!(layout.spider_col, v)
+    return
 end
 
 """
     qubit_loc(layout, v)
 
 Return the qubit number corresponding to the spider `v`.
 """
-qubit_loc(layout::ZXLayout{T}, v::T) where T = findfirst([(v in seq) for seq in layout.spider_seq])
+function qubit_loc(layout::ZXLayout{T}, v::T) where T
+    if haskey(layout.spider_q, v)
+        return T(layout.spider_q[v])
+    end
+    return
+end
+
+"""
+    column_loc(layout, v)
+
+Return the column number corresponding to the spider `v`.
+"""
+function column_loc(layout::ZXLayout{T}, v::T) where T
```

Comment:
no need to write it, just use `get(layout.spider_col, v, nothing)`

Reply:
I think it's necessary to add an API for this

Reply:
I mean

```suggestion
column_loc(layout::ZXLayout, v) = get(layout.spider_col, v, nothing)
```

just try to make things simpler